pkg/scheduler/api/job_info.go

```diff
@@ -719,21 +723,114 @@ func (ji *JobInfo) PendingBestEffortTaskNum() int32 {
 	return int32(count)
 }
 
+// FitFailedRoles returns the job roles' failed fit records
+func (ji *JobInfo) FitFailedRoles() map[string]struct{} {
+	failedRoles := map[string]struct{}{}
+	for tid := range ji.NodesFitErrors {
+		task := ji.Tasks[tid]
+		failedRoles[task.TaskSpec] = struct{}{}
+	}
+	return failedRoles
+}
+
+// TaskHasFitErrors checks if the task has fit errors and can continue try predicating
+func (ji *JobInfo) TaskHasFitErrors(task *TaskInfo) bool {
+	// if the task didn't set the spec key, should not use the cache
+	if len(task.TaskSpec) == 0 {
+		return false
+	}
+
+	_, exist := ji.FitFailedRoles()[task.TaskSpec]
+	return exist
+}
+
+// CheckJobNeedContinueAllocating checks whether it can continue on allocating for current job
+// when its one pod predicated failed, there are two cases to continue:
+//  1. job's total allocatable number meet its minAvailable(each task role has no independent minMember setting):
+//     because there are cases that some of the pods are not allocatable, but other pods are allocatable and
+//     the number of this kind pods can meet the gang-scheduling
+//  2. each task's allocable number meet its independent minAvailable
+//     this is for the case that each task role has its own independent minMember.
+//     eg, current role's pod has a failed predicating result but its allocated number has meet its minMember,
+//     the other roles' pods which have no failed predicating results can continue on
+//
+// performance analysis:
+//
+//	As the failed predicating role has been pre-checked when it was poped from queue,
+//	this function will only be called at most as the number of roles in this job.
+func (ji *JobInfo) CheckJobNeedContinueAllocating() bool {
```

Comment:
CheckJobNeedContinueAllocating() is used to handle the elastic job whose 's minavailable is not equal to task numbers. If  ji.MinAvailable == len(Tasks), it's better return at once to avoid the useless computing.

Reply:
CheckJobNeedContinueAllocating -> needContinueAllocating()

Reply:
Done.

Reply:
> If ji.MinAvailable == len(Tasks), it's better return at once to avoid the useless computing

Good catch. I have considered the performance effect, but forget this one.